src/software/new_geom/cubic_bezier_spline2d.h

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include "software/new_geom/bezier_curve2d.h"
+#include "software/new_geom/ray.h"
+#include "software/new_geom/spline2d.h"
+
+/**
+ * A Cubic Bezier spline is a C2 continuous spline that is formed by a series of bezier
+ * curves.
+ *
+ * Some useful resources:
+ * https://www.ibiblio.org/e-notes/Splines/Intro.htm
+ * https://www.ibiblio.org/e-notes/Splines/b-int.html
+ */
+class CubicBezierSpline2d : public Spline2d
+{
+   public:
+    CubicBezierSpline2d() = delete;
+
+    /**
+     * Create a CubicBezierSpline2d that:
+     *  - starts from `start` and is tangent to `start_vector` at the start
+     *  - ends at `end` and is tangent to `end_vector` at the end
+     * @param start_point The first point on the spline
+     * @param start_vector The tangent vector at the start of the spline
+     * @param end_point The last point on the spline
+     * @param end_vector The tangent vector at the end of the spline.
+     *                   NOTE: this points "back towards" the start of the spline, instead
+     *                         of "off the end" of the spline. ie. this vector is
+     *                         "symmetric" with `start_vector`
```

Comment:
I kinda understand what you're trying to say, but it's confusing to think about a vector as ending at a point rather than starting at a point. Can you link a source or make an ASCII diagram for how this works?

Reply:
Sure, added this:
```
     *    ** EXAMPLE: **
     *
     *                           XXXXXXXXX
     *        start_vector  XXXXX         XXXXX   end_vector
     *              ^     XXX                 XXX     ^
     *              |    XX                     XX    |
     *              |   XX                       XX   |
     *              |  XX                         XX  |
     *              |  X                           X  |
     *              | XX                           XX |
     *              | X                             X |
     *              | X                             X |
     *              |XX                             XX|
     *              |X                               X|
     *              |X                               X|
     *              |X                               X|
     *       start  +                                 +  end
```

Reply:
very nice ascii art :+1: